examples/move/scratch_off/sources/game.move

```diff
@@ -0,0 +1,549 @@
+        &store.player_metadata
+    }
+
+    public fun get_target_player_metadata(player_metadata: &Table<address, Metadata>, target_address: address): &Metadata {
+        table::borrow(player_metadata, target_address)
+    }
+
+    public fun get_player_metadata_mut(player_metadata: &mut Table<address, Metadata>, target_address: address): &mut Metadata {
+        table::borrow_mut(player_metadata, target_address)
+    }
+
+    public fun original_ticket_count<Asset>(store: &ConvenienceStore<Asset>): u64 {
+        store.original_ticket_count
+    }
+
+    public fun tickets_left<Asset>(store: &ConvenienceStore<Asset>): u64 {
+        store.original_ticket_count - store.tickets_evaluated
+    }
+
+    public fun tickets_issued<Asset>(store: &ConvenienceStore<Asset>): u64 {
+        store.tickets_issued
+    }
+
+    public fun prize_pool_balance<Asset>(store: &ConvenienceStore<Asset>): u64 {
+        balance::value(&store.prize_pool)
+    }
+
+    /// OTW to claim Publisher object, in order to create Display.
+    struct GAME has drop {}
+
+    /// We claim the cap for updating display
+    fun init(otw: GAME, ctx: &mut TxContext){
+        package::claim_and_keep(otw, ctx);
+    }
+
+    /// Initializes the store with all of the lottery tickets.
+    /// We allow the user of the store.
+    /// We purposely design the convenience store to be an owned object so that we
+    /// don't need to make this in a shared format.
+    public fun open_store<Asset>(
+        coin: Coin<Asset>, 
+        number_of_prizes: vector<u64>,
+        value_of_prizes: vector<u64>,
+        public_key: vector<u8>,
+        max_players_in_leaderboard: u64,
+        ctx: &mut TxContext
+    ): Coin<Asset> {
+        let number_of_prizes_len = vector::length(&number_of_prizes);
+        let value_of_prizes_len = vector::length(&value_of_prizes);
+        assert!(number_of_prizes_len == value_of_prizes_len, EInvalidInputs);
+
+        let winning_tickets = vector<PrizeStruct>[];
+        let idx = 0;
+        let prize_pool = balance::zero<Asset>();
+        let winning_ticket_count = 0;
+
+        while (idx < number_of_prizes_len) {
+            let target_prize_amount = vector::pop_back(&mut number_of_prizes);
+            let target_prize_value = vector::pop_back(&mut value_of_prizes);
+            vector::push_back(&mut winning_tickets, PrizeStruct {
+                ticket_amount: target_prize_amount,
+                prize_value: target_prize_value
+            });
+
+            // prize_amount * prize_value to get the total we need.
+            // Note that this is a u64 * u64 so we could have overflow but for the purpose
+            // of this smart contract we do not need to consider this.
+            let target_amount = target_prize_amount * target_prize_value;
+            // Pull the required balance from the coin and stuff it into a balance.
+            balance::join(&mut prize_pool, coin::into_balance(coin::split(&mut coin, target_amount, ctx)));
```

Comment:
Adding on to the previous comment, I'd probably just merge the full Coin on the contract, and just do a total check that `coin::value(coin) == required_funds`, instead of splitting the coin for each prize

Reply:
Updated to this